entities/src/main/java/org/odk/collect/entities/javarosa/filter/LocalEntitiesFilterStrategy.kt

```diff
@@ -36,32 +38,81 @@ class LocalEntitiesFilterStrategy(entitiesRepository: EntitiesRepository) :
+    ): Query? {
+        return when (predicate) {
+            is XPathBoolExpr -> xPathBoolExprToQuery(predicate, sourceInstance, evaluationContext)
+            is XPathEqExpr -> xPathEqExprToQuery(predicate, sourceInstance, evaluationContext)
+            else -> null
+        }
+    }
+
+    private fun xPathBoolExprToQuery(
+        predicate: XPathBoolExpr,
+        sourceInstance: DataInstance<*>,
+        evaluationContext: EvaluationContext,
+    ): Query? {
+        val queryA = xPathExpressionToQuery(predicate.a, sourceInstance, evaluationContext)
+        val queryB = xPathExpressionToQuery(predicate.b, sourceInstance, evaluationContext)
+
+        return if (queryA != null && queryB != null) {
+            if (predicate.op == XPathBoolExpr.AND) {
+                Query.And(queryA, queryB)
+            } else {
+                Query.Or(queryA, queryB)
+            }
+        } else {
+            null
+        }
+    }
+
+    private fun xPathEqExprToQuery(
+        predicate: XPathEqExpr,
+        sourceInstance: DataInstance<*>,
+        evaluationContext: EvaluationContext,
+    ): Query? {
         val candidate = CompareToNodeExpression.parse(predicate)
-        return when (val original = candidate?.original) {
-            is XPathEqExpr -> {
-                if (original.isEqual) {
-                    val child = candidate.nodeSide.steps[0].name.name
-                    val value = candidate.evalContextSide(sourceInstance, evaluationContext)
 
-                    val results = instanceAdapter.queryEq(
-                        sourceInstance.instanceId,
-                        child,
-                        value as String
-                    )
+        return if (candidate != null) {
+            val child = candidate.nodeSide.steps[0].name.name
+            val value = candidate.evalContextSide(sourceInstance, evaluationContext) as String
 
-                    sourceInstance.replacePartialElements(results)
-                    results.map {
-                        it.parent = sourceInstance.root
-                        it.ref
-                    }
-                } else {
-                    next.get()
-                }
+            if (predicate.isEqual) {
+                Query.Eq(child, value)
+            } else {
+                Query.NotEq(child, value)
             }
+        } else {
+            null
+        }
+    }
 
-            else -> {
-                next.get()
+    private fun queryToTreeReferences(
+        query: Query?,
+        sourceInstance: DataInstance<*>,
+        next: Supplier<MutableList<TreeReference>>
+    ): List<TreeReference> {
+        return if (query != null) {
+            val results = instanceAdapter.query(sourceInstance.instanceId, query)
+            sourceInstance.replacePartialElements(results)
+            results.map {
+                it.parent = sourceInstance.root
+                it.ref
             }
+        } else {
+            next.get()
```

Comment:
It doesn't look like this path gets tested (from running with coverage). We need cases for non-optimized expressions (like functions) and for expressions where part of them isn't supported (for the two `else` clauses that return `null` in `xPathBoolExprToQuery` and `xPathEqExprToQuery`).

Reply:
Actually, this code was redundant. We have a null check before calling this method and there is one test that covers this case: `works correctly with non eq name expressions`